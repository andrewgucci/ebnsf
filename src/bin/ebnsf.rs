use ebnsf::parse_ebnf;

use std::path::PathBuf;

fn main() {
    let ebnf = std::fs::read_to_string("test/bnf.ebnf").unwrap();

    let diagram = parse_ebnf(&ebnf).unwrap();

    let output = PathBuf::from("test.svg");

    std::fs::write(&output, diagram.to_string().into_bytes()).unwrap();
}
