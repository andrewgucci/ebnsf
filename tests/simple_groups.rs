use ebnsf::parse_ebnf;

mod common;
use common::save_svg;

#[test]
fn simple_groups() {
    let rule = "\
<rule> ::= <foo> (<bar> \"|\" <bish> <bosh>)*
<rule> ::= <foo> | (<bish> | <bosh>)
<rule> ::= <foo> | (<bish> | <bosh>)?
<rule> ::= <foo> | (<bish> | <bosh>)+
<rule> ::= <foo> | (<bish> | <bosh>)*
<rule> ::= (<foo> | (<bish> | <bosh>)*)?
";

    let svg = parse_ebnf(rule).expect("valid grammar").to_string();

    save_svg("simple_groups", &svg);

    insta::assert_snapshot!("simple_groups", svg);
}

#[test]
fn multiline_groups() {
    let rule = "\
<rule> ::= <foo>
        | (<foo> | <bar>)
        | (
                <fizz>
              | (\"foo\" | \"bar\")
          )
";

    let svg = parse_ebnf(rule).expect("valid grammar").to_string();

    save_svg("multiline_groups", &svg);

    insta::assert_snapshot!("multiline_groups", svg);
}
