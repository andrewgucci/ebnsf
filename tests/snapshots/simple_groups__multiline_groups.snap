---
source: tests/simple_groups.rs
expression: svg
---
<svg class="railroad" viewBox="0 0 306 207" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<style type="text/css">
svg.railroad {
background-color: hsl(30, 20%, 95%);
background-size: 15px 15px;
background-image: linear-gradient(to right, rgba(30, 30, 30, .05) 1px, transparent 1px),
		  linear-gradient(to bottom, rgba(30, 30, 30, .05) 1px, transparent 1px);
}

svg.railroad rect.railroad_canvas {
stroke-width: 0px;
fill: none;
}

svg.railroad path {
stroke-width: 3px;
stroke: black;
fill: none;
}

svg.railroad .debug {
stroke-width: 1px;
stroke: red;
}

svg.railroad text {
font: 14px monospace;
text-anchor: middle;
}

svg.railroad .nonterminal text {
font-weight: bold;
}

svg.railroad text.comment {
font: italic 12px monospace;
}

svg.railroad rect {
stroke-width: 3px;
stroke: black;
fill:hsl(-290, 70%, 90%);
}

svg.railroad g.labeledbox > rect {
stroke-width: 1px;
stroke: grey;
stroke-dasharray: 5px;
fill:rgba(90, 90, 150, .1);
}
</style>
<rect class="railroad_canvas" height="100%" width="100%"/>
<g class="verticalgrid">
<g class="sequence">
<path d=" M 10 21 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="54" y="26">
rule</text>
<g class="choice">
<path d=" M 83 21 h 24 m 44 0 h 120 m -57 0 l -5 -5 m 0 10 l 5 -5 m 57 0"/>
<g class="sequence">
<g class="nonterminal">
<rect height="22" width="44" x="107" y="10"/>
<text x="129" y="26">
foo</text>
</g>
</g>
<path d=" M 83 21 a 12 12 0 0 1 12 12 v 9 m 164 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 95 42 v 66 m 0 -30 l -5 -5 m 10 0 l -5 5 m 0 30 m 164 0 v -66 m 0 36 l -5 5 m 10 0 l -5 -5 m 0 -36"/>
<path d=" M 95 42 v 0 a 12 12 0 0 0 12 12 m 92 0 h 48 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="choice">
<path d=" M 107 54 h 24 m 44 0 h 24"/>
<g class="sequence">
<g class="nonterminal">
<rect height="22" width="44" x="131" y="43"/>
<text x="153" y="59">
foo</text>
</g>
</g>
<path d=" M 107 54 a 12 12 0 0 1 12 12 v 9 m 68 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 119 75 v 0 a 12 12 0 0 0 12 12 m 44 0 h 0 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="nonterminal">
<rect height="22" width="44" x="131" y="76"/>
<text x="153" y="92">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 95 108 v 0 a 12 12 0 0 0 12 12 m 140 0 h 0 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="choice">
<path d=" M 107 120 h 24 m 52 0 h 64 m -29 0 l -5 -5 m 0 10 l 5 -5 m 29 0"/>
<g class="sequence">
<g class="nonterminal">
<rect height="22" width="52" x="131" y="109"/>
<text x="157" y="125">
fizz</text>
</g>
</g>
<path d=" M 107 120 a 12 12 0 0 1 12 12 v 9 m 116 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 119 141 v 0 a 12 12 0 0 0 12 12 m 92 0 h 0 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="choice">
<path d=" M 131 153 h 24 m 44 0 h 24"/>
<g class="sequence">
<g class="terminal">
<rect height="22" rx="10" ry="10" width="44" x="155" y="142"/>
<text x="177" y="158">
foo</text>
</g>
</g>
<path d=" M 131 153 a 12 12 0 0 1 12 12 v 9 m 68 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 143 174 v 0 a 12 12 0 0 0 12 12 m 44 0 h 0 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="terminal">
<rect height="22" rx="10" ry="10" width="44" x="155" y="175"/>
<text x="177" y="191">
bar</text>
</g>
</g>
</g>
</g>
</g>
</g>
</g>
<path d=" M 73 21 h 10"/>
</g>
<path d=" M 281 21 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 21 h 10"/>
<path d=" M 271 21 h 10"/>
</g>
</g>
</svg>
